### ISSUE TITLE
Token Expiration/Renew causing race condition with okta-angular

### ISSUE THREAD

**Issue by polklabs** [okta/okta-auth-js#732]
Not sure if this should be posted here or in okta-angular.

The auto renew of tokens is causing a race condition in conjunction with okta-angular's **OktaAuthGuard**

When the `idToken` or `accessToken` expire the **TokenManager** fires an **EVENT_EXPIRED** for each of them. The **TokenService/TokenManager** then attempt to renew them, all good so far. During this renew process the token is removed from storage and the new token is added back in. `Add()` and `Remove()` both emit events that the AuthStateManager listens for, triggering the `updateAuthState()`.
The problem occurs because the okta-angular **OktaAuthGuard** is listening for that state change and if it sees that the authState is no longer authenticated it calls `onAuthRequired()`, which can happen if the **AuthStateManager** emits an event in the middle of a token renewal.

What I believe is happening is that depending on the expiration timing, `updateAuthState()` can be triggered during `Remove()` for the second token causing that token to be null and therefore isAuthenticated=false.

Without changing anything in my code, I was able to trigger both the expected renewal process and the renewal with the issue. The ONLY thing I did between these tests was refresh the page. I added the following code to capture the issue

```
     this.oktaAuth.tokenManager.on('expired', (key) => {
        console.log(`Event 'expired': ${key}`);
      });

      this.oktaAuth.authStateManager.subscribe((authState) => {
        console.log(authState);
      });

      this.oktaAuth.tokenManager.on('added', () => {
        console.log('Token ADDED!');
      });

      this.oktaAuth.tokenManager.on('removed', () => {
        console.log('Token REMOVED!');
      });
```

### Working as expected

> Event 'expired': accessToken
> Token REMOVED!
> Token ADDED!
> AuthState: {accessToken: {…}, idToken: {…}, refreshToken: undefined, isPending: false, isAuthenticated: true}
> Event 'expired': idToken
> Token REMOVED!
> Token ADDED!
> AuthState: {accessToken: {…}, idToken: {…}, refreshToken: undefined, isPending: false, isAuthenticated: true}

### Not working as expected

> Event 'expired': accessToken
> Event 'expired': idToken
> Token REMOVED!
> Token ADDED!
> AuthState: {accessToken: {…}, idToken: null, refreshToken: undefined, isPending: true, isAuthenticated: false}
> Token REMOVED!
> Token ADDED!
> AuthState: {accessToken: {…}, idToken: {…}, refreshToken: undefined, isPending: false, isAuthenticated: true}

As you can see in the second output, the `idToken` is **null** when it is in the process of being renewed. **OktaAuthGuard** sees that same auth state and thinks that it needs to call `handleLogin()` -> `onAuthRequired()`.

It appears that there needs to be some sort of small delay between the expirations or wait until both tokens have been renewed before emitting the authState.

**Comment by bryanapellanes-okta** [comment]
@polklabs,
Thanks for reaching out! I'll bring this to the attention of our Javascript developers.  Someone will follow up here with more.

Thanks for using Okta!

**Comment by bryanapellanes-okta** [comment]
Internal Ref: OKTA-395165

**Comment by aarongranick-okta** [comment]
@polklabs Thank you for bringing this to our attention! I believe this is a bug, likely happening here: https://github.com/okta/okta-auth-js/blob/4.9/lib/AuthStateManager.ts#L155

I think this issue has been fixed in 5.0, unfortunately okta-angular has not been updated yet to support the changes. (but this is coming very soon)

We will make a fix for auth-js 4.9

**Comment by polklabs** [comment]
@aarongranick-okta That does look like the source of the bug and matches the behavior that I'm seeing.

**Comment by aarongranick-okta** [comment]
@polklabs A fix has been pushed as `@okta/okta-auth-js` version 4.9.1

**Comment by aarongranick-okta** [comment]
@polklabs Is it working as expected using auth-js version 4.9.1 ?

**Comment by polklabs** [comment]
@aarongranick-okta Yes it's working, I have not encountered the issue again since updating to 4.9.1

**Comment by jaredperreault-okta** [comment]
@prashantagrawalcrowe Please file a new issue, this issue was closed almost 3 years ago. Please provide as many details as you have in the new issue. Thanks
